documentation/sphinx/source/rangelock.rst

```diff
@@ -0,0 +1,122 @@
+
+Get locked ranges within the input range
+
+``ACTOR Future<std::vector<KeyRange>> getReadLockOnRange(Database cx, KeyRange range);``
+
+Register a rangeLock owner to database metadata.
+
+``ACTOR Future<Void> registerRangeLockOwner(Database cx, std::string uniqueId, std::string description);``
+
+Remove an owner from the database metadata
+
+``ACTOR Future<Void> removeRangeLockOwner(Database cx, std::string uniqueId);``
+
+Get all registered rangeLock owner
+
+``ACTOR Future<std::vector<RangeLockOwner>> getAllRangeLockOwners(Database cx);``
+
+Get the owner of a rangeLock
+
+``ACTOR Future<Optional<RangeLockOwner>> getRangeLockOwner(Database cx, std::string uniqueId);``
+
+
+Example usage
+-------------
+Turn on/off user traffic for bulk load based on range lock. Note that ideally, we want to use the write lock to achieve this, now we use the read lock as a temporarily solution.
+
+``ACTOR Future<Void> turnOffUserWriteTrafficForBulkLoad(Transaction* tr, KeyRange range);``
+
+``ACTOR Future<Void> turnOnUserWriteTrafficForBulkLoad(Transaction* tr, KeyRange range);``
+
+
+Per-Range Lock Design
+=====================
+The per-range lock information is persisted in ``\xff/rangeLock/`` system key space.
+Users specify ranges to lock in ``\xff/rangeLock/`` system key space via a transaction. 
+The range lock can be only within user key space, aka ``"" ~ \xff``.
+The value of the key space is either empty or setting as a set of locks. 
+If a range having a ReadLock, the cluster will reject any mutation within the range. 
+If the range value is empty, the cluster does not reject the traffic as the range is unlocked.
+
+The per-range lock API issues transactions to update the ``\xff/rangeLock/`` system metadata. 
+When the mutation arrives at commit proxies, the commit proxies keep the in-memory key range map of range locking status, 
+and they update the map when applying the metadata mutation. 
+When following normal mutation arrives at the commit proxy, the commit proxy checks the range locking status in the in-memory map.
+If the mutation is within a locked range, the commit proxy rejects the mutation.
+This process happens in the phase of postResolution, where the commit proxy scans all transactions. 
+In addition, the lockRange mutation is also persisted to txnStateStore. When recovery, a new commit proxy will rebuild the in-memory lockRange 
+map based on the status persisted in txnStateStore.
+
+When a locking mutation arrives at a commit proxy, 
+the lock takes effect on all mutations that arrive at any commit proxy after this locking mutation. 
+Note that it is possible that this locking mutation is batched with other mutations with the same commit version. 
+The locking mutation takes effect on all mutations after the locking mutation in the same batch.
+To achieve this, the locking transaction adds a write_conflict_range on the entire user key space (``"" ~ \xff``).
```

Comment:
is this correct? I thought the write conflict range is set to be the locked range.

Reply:
Nice catch. I think you are correct. We only need to add the locked range.

Reply:
OK. Can you update the doc here?